content/lessons/advanced-topics/dynamodb-streams.md

```diff
@@ -7,4 +7,150 @@ date: "01/01/2018"
+                  "S":"This is the new attribute data"
+               },
+               "MyPartitionKey":{
+                  "S":"899ff269-47b4-4aca-aac1-a7da2c528512"
+               }
+            },
+            "OldImage":{
+               "Message":{
+                  "S":"This is the old attribute data"
+               },
+               "MyPartitionKey":{
+                  "S":"899ff269-47b4-4aca-aac1-a7da2c528512"
+               }
+            }
+         }
+      },
+      {
+         "eventID":"3",
+         "eventName":"REMOVE",
+         "dynamodb":{
+            "Keys":{
+               "MyPartitionKey":{
+                  "S":"14cb3e02-d5f4-4c4a-b7d5-2811c7055552"
+               }
+            },
+            "OldImage":{
+               "MyData":{
+                  "S":"Some data"
+               },
+               "MyPartitionKey":{
+                  "S":"14cb3e02-d5f4-4c4a-b7d5-2811c7055552"
+               }
+            }
+         }
+      }
+   ]
+}
+```
+
+This is an array of three items from a stream. The lambda code now can react to whatever attribute is relevant.
+
+**eventID** is the unique ID of the stream item. Use this for logging, tracing, de-duplication (think idempotency!)
+
+**eventName** tells you wheter an item was added, changed or removed, so you can react accordingly. Be aware that for _NEW_AND_OLD_IMAGES_ type streams INSERT events only have the NewImage element, REMOVE events only have the OldImage element.
+
+With this information, your lambda function can react to changes on the stream, like processing data right away, or sending to downstream services.
+
+## Aggregations ##
+
+Another common use case for DynamoDB stream processing is aggregating data as the table changes.
+
+As DynamoDB is not designed to deliver search engine capabilities and as such is not built to let you aggregate table data natively.
+
+A good approach though, is to use the stream of a DynamoDB table to listen to changes as they come through and in the processing lambda function to whatever aggregation is needed.
+
+This could look roughly like this:
+
+```
+var docClient = new AWS.DynamoDB.DocumentClient();
+
+var params = {
+    TableName:"...",
+    Key: {
+        "MyPartitionKey" : "AggSumOfSomething"
+    },
+    UpdateExpression: "add TheSum = :countToAdd",
+    ExpressionAttributeValues:{
+        ":countToAdd":value
+    }
+};
+
+docClient.update(params,callback);
+```
+
+In this case, you would have a dedicated item in your table with a partition key of _AggSumOfSomething_ that you would use to keep track of attribute values in your table items. Of course, if _value_ is negative it would be subtracted from the aggregate.
+
+**Careful:** Don't create yourself a hot partition by firing to much at this single aggregate item. If you have to, look up "write sharding". Also, obviously the aggregate is eventual consistent as you probably can't guarantee live processing of you stream at all time.
```

Comment:
You could add a link to the aws docs or https://www.dynamodbguide.com/leaderboard-write-sharding, where write sharding is explained